### PR TITLE
Alloy Transmuter Support

### PR DESCRIPTION
**Pull request by iboss-ptk**
- Add support for alloy transmuter
- Use cw2 info to determine the pool type
- Pass along `CosmWasmPoolModel` with information queried from contract to sqs
- Fix CosmWasm pool conversion failed on streaming update due to missing wasmkeeper

sqs tag for these changes block https://github.com/osmosis-labs/osmosis/pull/8291

### CODE REVIEW

File: sqsdomain/pools.go
Comment (p0mvn) [inline sqsdomain/pools.go]: nit: to make it clearer for any future new contributors
```suggestion
	GetCosmWasmPoolModel() (*CWPoolModel, error)
```

File: sqs.log
Comment (p0mvn) [inline sqs.log]: nit: do you mind deleting this file before final review? 🙏

File: sqsdomain/pools.go
Comment (p0mvn) [inline sqsdomain/pools.go]: Can we instead define `*CWPoolModel` on `SQSModel`?

That way, we can avoid defining concrete CosmWasm getter and setter APIs on the generalized `PoolI` interface, helping to keep it lean.

Pet this suggestion, we would be able to use `GetSQSPoolModel().CWPoolModel`

**Review by p0mvn — COMMENTED** [review]
Just did a quick skim without any deep review for now. Looks like everything is going well and on the right track. Great work!

File: sqs.log
Comment (iboss-ptk) [reply sqs.log]: noted

File: sqsdomain/pools.go
Comment (iboss-ptk) [reply sqsdomain/pools.go]: sounds good! will do that

File: sqsdomain/cw_pool.go
Comment (p0mvn) [inline sqsdomain/cw_pool.go]: Would it make sense to set this via config?

File: sqsdomain/cw_pool.go
Comment (p0mvn) [inline sqsdomain/cw_pool.go]: nit: could we add godocs to all types. I'm sure we would benefit from these in the future

File: router/usecase/router.go
Comment (p0mvn) [inline router/usecase/router.go]: This log does not make sense to me to be in this location

File: router/usecase/pools/pool_factory.go
Comment (p0mvn) [inline router/usecase/pools/pool_factory.go]: I think we should still have the ability to control whether we support an alloyed transmitter by having a config for them somewhere [here](https://github.com/osmosis-labs/sqs/blob/7f9fdda008ac7dd6ab599caec64622d375543d44/config.json#L20).

In previous cw pool releases, we had several unexpected issues arise so controlling what is supported by our router proved to be useful.

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: Could you link some docs to https://github.com/osmosis-labs/transmuter that explain normalization factors please?

I think that future contributors would benefit from this

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go:176]: Can we handle unexpected truncation to zero here?

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go:62]: Could we add a test for truncation in division occurring here?

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: Would be great to unit test each sub-method on the calc path individually - the happy paths and major error/edge cases 🙏

**Review by p0mvn — COMMENTED** [review]
Great work! Left some comments

**Comment by p0mvn** [comment]
Also, would be great to document somewhere the logic around how `CWPoolModel` relies on the contract info and how this is different from transmuter and generalized cw pools. Maybe we could start a doc under `docs/architecture/CW_POOLS.MD`?

File: sqsdomain/cw_pool.go
Comment (iboss-ptk) [reply sqsdomain/cw_pool.go]: I don't think so as this function is referenced by ingester codebase as well. Also this information will not be varied from environment to environment.

**Comment by p0mvn** [comment]
@iboss-ptk, please ping me once ready for another round of 👀

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go:176]: Not sure which part would you expect truncation to happen? Are you referring to high `tokenInNormFactor` that might turn result into 0?  What would an expected behavior in such case would you like it to be?

File: domain/router.go
Comment (p0mvn) [inline domain/router.go]: side nit: could you format this to be on the same indent with other fields?

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go:176]: Thinking more about this - I think it is fine if we truncate. In that case, the value is probably small to the extent of being insignificant

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go:176]: Does it make sense to add a division by zero check?

Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: Could we also maximize the usage of mutative (suffixed with `Mut`) versions of the math operations for performance reasons?

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go:62]: I think the test case for this is still applicable.

We can choose the denominator large enough and numerator small enough to take the final result to zero despite the starting inputs being non-zero

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: nit: can we define this at the top of the test fixture to avoid reinitializing for every single one

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: I understand why we're choosing to use this boilerplate but could we either add comments or encapsulate this in a separate function for readability?

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go:200]: Can we add a test case where the normalization factor is zero to prompt a division by zero and panic in the final calculation?

We should probably handle that gracefully and return an error instead

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: We could also reuse the initialisation helped across different test cases

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go:124]: I like how you added a separate test for a lower-level calc method but could we also unit-test this one?

Seems like a few things are happening here, and we would definitely benefit from strict coverage

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: noted

Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go:176]: `normalization_factor` should never be 0 and it's validated in contract. What would fail gracefully looks like for you? explicitly return error?

Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: It seems that `Mut` variants does not have `MulInt` and `QuoInt` but only `MulMut` and `QuoMut` which takes Dec and will loose precision if converting those number to Dec. Do we want that for performance?

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go:124]: There are already unit-test cases for this function.

File: config.json
Comment (p0mvn) [inline config.json]: Can you please update the following and merge them before this PR is merged?

Template: https://github.com/osmosis-labs/infrastructure/blob/55a5946330250d38047d26f8c4865474436db9d3/environments/sqs-osmosis-zone/ansible/roles/prepare_local_folder/templates/sqs/config.json.j2

Configs:
- https://github.com/osmosis-labs/infrastructure/blob/main/environments/sqs-osmosis-zone/environments/stage/ansible/vars/sqs.yaml
- https://github.com/osmosis-labs/infrastructure/blob/main/environments/sqs-osmosis-zone/environments/prod/ansible/vars/sqs.yaml
- https://github.com/osmosis-labs/infrastructure/blob/main/environments/sqs-osmosis-zone/environments/testnet/ansible/vars/sqs.yaml

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: nit: can we define constants for the same denoms. Would make test updates simpler in the future

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: re: the previous comment about constants. I think by having them referenced via shared variables, noting such dependent setup logic would be simpler since we could search variables by references

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: nit: ditto for vars for int values to simplify future updates and reference checking.

Applies to other tests as well.

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: In Go, it is idiomatic to return a nil value rather than an initialized zero when an error is returned.

Would it make sense to return an uninitialized `osmomath.BigDec{}` here for consistency

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (p0mvn) [inline router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: Could we do `tokenOutAmt.Dec().TruncateInt()` to simplify the conversion?

File: sqsdomain/cw_pool.go
Comment (p0mvn) [inline sqsdomain/cw_pool.go:81]: Can these be regular Go ints?

If so, we would save on math operations and serialization overhead while also improving readability from not having to do `osmomath.Int` to `osmomath.BigDec` conversions.

**Review by p0mvn — COMMENTED** [review]
This already looks good to me. Nice work!

However, I would like to run the tests before approving. Will come back to this.

Btw please make PRs from this repo directly (not fork) - it would simplify the review process in the future.

I will add you to the repo

File: sqsdomain/cw_pool.go
Comment (p0mvn) [inline sqsdomain/cw_pool.go]: What if the contract is migrated? We would then require code updates and redeployment?

Please consider removing this hard-coded dependency here and on the node side.

The current code ID model allows us to simply update the config without performing any code changes

Comment (p0mvn) [reply sqsdomain/cw_pool.go]: Please consider going back to the code ID config when determining if a pool is an alloy transmuter within SQS. 

To achieve this on the Osmosis node side, we can continue utilizing the CW info but only match the name e.g. `crates.io:transmuter-alloy` (no version). This way, if the version is updated, it will still be auto-propagated from Osmosis node. However, give SQS the flexibility of handling it via standard means.

The proposed code ID approach would ensure compatibility with how other cw pool types are handled while also allowing the flexibility of only updating configs without code changes if migration is done.

With the current approach, we still need a boolean config but also hardcode the version here which I think is suboptimal for operations.

Please let me know what you think

File: sqsdomain/cw_pool.go
Comment (iboss-ptk) [reply sqsdomain/cw_pool.go:81]: It can't, normalization factor query response is represented as string due to potential misprecision in javascript client. So the responded value is not just an int.

File: sqsdomain/cw_pool.go
Comment (iboss-ptk) [reply sqsdomain/cw_pool.go]: noted

Comment (iboss-ptk) [reply sqsdomain/cw_pool.go]: but then, this logic required on the ingester side, do we also have config on that side, I didn't see any. If there is any update required, both side needs to be effected.

Comment (iboss-ptk) [reply sqsdomain/cw_pool.go]: I think versions are more reliable than code ids due to it's characteristics that it does not change from env to env. That's why I think it's ok to hard code it and have it defined once (in sqs repo). Even though the contract gets migrated and it requires no change on sqs. It will still work. These infos are bound to conversion logic anyway.

For versioning matching, we can use semver matching to check on which pool version we are operating on as well to increase range of compatability. For example, for alloy transmuter it can be ">=3.0.0" for now.

Other cosmwasmpool can also adopt this approach (I didn't update then in this PR since I would love to push this out earlier to test it with USDT alloyed, it requires a bit more regression testing).

Bottom line is that, we can't avoid code changes if the logics for calculating token out changes. We can be lax on the config (eg. ">=3.0.0" and no restriction on the upper bound) until there is a version that requires new calculation. We can have sqs disregard newer versions via config if that's a desired property of the system to control incompatible pool type to not get calculated.

On another point, "matching name only" does not work well IMO. Since there might be changes in the pool transformation logic in the future version, which means we have to change contract name again.

File: config.json
Comment (iboss-ptk) [reply config.json]: I can't access `infrastructure` repo

File: domain/router.go
Comment (iboss-ptk) [reply domain/router.go]: done

File: router/usecase/pools/routable_cw_alloy_transmuter_pool.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool.go]: done

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: done

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: done

File: router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go
Comment (iboss-ptk) [reply router/usecase/pools/routable_cw_alloy_transmuter_pool_test.go]: done

File: sqsdomain/cw_pool.go
Comment (p0mvn) [reply sqsdomain/cw_pool.go]: For visibility, this is Boss's reply (did not get submitted publicly) that I will address in a separate message

image

Comment (p0mvn) [reply sqsdomain/cw_pool.go]: I think that we should discuss this decision via a separate ADR to avoid blocking this line of work.

I would like to avoid merging partial progress with a new paradigm that has trade-offs. If we want to change how CW pools are handled, let's aim to change this for all pools at once rather than introducing custom handling for the alloyed pool but keeping everything else via code IDs.

What I would like to avoid is us never coming back to this and then having to deal with operational challenges stemming from the custom logic.

In terms of versions, I don't understand why we need to track these hard-coded versions in general when contract names should suffice. We should establish a rule that all future versions of a cw pool model struct transferred between node and sqs are backward compatible. This removes the need for version matching on the node side, removing the need to either have a config (node-side) or hard code it in `sqsdomain`.

I'm supportive of using versions in the SQS config rather than hardcoding. But again, I propose making that decision in a separate ADR and/or issue to avoid overly long debates on the otherwise complete work here.

Again, my main counter-argument to the proposed change in this PR is the following:
- Operational challenges stemming from hardcoding versions on both osmosis and sqs side
- Applying an operational change to one cw pool only, causing inconsistencies in handling
- The trade-offs are not fully analyzed before making a change that impacts service operations

However, I think that we should explore the version idea in SQS configs in general. That is an interesting proposal.

@iboss-ptk please let me know what you think

Comment (iboss-ptk) [reply sqsdomain/cw_pool.go]: Fair point on consistency.

I will revert back the config to code ids and and have a separated ADR for versioning we discussed to unblock things. That being said. Since old transmuter is not forward compatible, it still need minimum version `>=3.0.0` hardcoded. Because even though it is a non-issue on the sqs side in case there is no minimum version in place, ingester needs this information to do pool transformation properly as old transmuter does not have those queries we use for v3.0.0. 

And current transmuter v3 on mainnet is already `crates.io:transmuter` `v3.0.0` so this can't be changed to `crates.io:transmuter-alloy` unless having a migration (and I don't think that make much sense anyway since it is still a newer version of the same thing).


To summarize:
- I don't think we can get rid of any hardcoded version except making more lax and keep backward compatibility. Which might or might not make sense depending on

File: sqsdomain/cw_pool.go
Comment (p0mvn) [inline sqsdomain/cw_pool.go:62]: I like this approach. It is now extensible without having to make code changes to both, the `osmosis` node side and `sqs` for minor migrations.

Can we also add constraint of semver being less than <4.0.0 so that, if there are any future versions, those are auto-accounted for without more code changes?

File: sqsdomain/cw_pool.go
Comment (p0mvn) [reply sqsdomain/cw_pool.go]: Thanks

Per [this comment](https://github.com/osmosis-labs/sqs/pull/245#discussion_r1625856060), the latest solution is more extensible.

If we keep it as a convention that all alloyed transmuters stay within the same major version and add a relevant semver constraint (`<4.0.0`), we solve all future extensibility problems.

In the original approach, due to semver being hardcoded directly, we would have to make code changes and dependency updates to all of `sqsdomain`, `osmosis` and `sqs` if a migration was done.

Now, migrations within the same major are auto-accounted for

File: config-testnet.json
Comment (p0mvn) [inline config-testnet.json]:
```suggestion
        ]
```

File: docs/architecture/COSMWASM_POOLS.MD
Comment (p0mvn) [inline docs/architecture/COSMWASM_POOLS.MD:1]: Do you mind adding a paragraph about semver and the general new version handling we arrived at with the cw pools? For example, "alloyed transmuter versions are expected to exist within the same major version i.e >=3.0.0 and <4.0.0"

**Review by p0mvn — APPROVED** [review]

File: config-testnet.json
Comment (p0mvn) [reply config-testnet.json]: comitting nit to avoid having you wait on CI

**Comment by p0mvn** [comment]
@iboss-ptk thanks for addressing all operational concerns. Great work.

I pre-approved the PR but the following things remain:
- Update `infrastructure` repository with the new config values: https://github.com/osmosis-labs/infrastructure
  * Please do no merge before this is done because our stage deployments will break
  * You should have access. If still not, please ask in #eng-platform-support
- Add docs about semver handling

File: sqsdomain/cw_pool.go
Comment (iboss-ptk) [reply sqsdomain/cw_pool.go:62]: At least for now that I'm aware of, 4.0.0 will still be compatible, so I would avoid constraining that if you don't want code change. Until there is a version the requires changes in logic, I think leave this lax constraints would be easier to operate.

Comment (p0mvn) [reply sqsdomain/cw_pool.go:62]: I think we should have a standard where a separate major means a breaking change.

If `4.0.0` is compatible, can it be `3.1.0` instead?

Comment (iboss-ptk) [reply sqsdomain/cw_pool.go:62]: It's breaking other parts. For 4.0.0 we are planning to remove some entrypoints, putting more constraints on denom, but just does not effect sqs.

Comment (p0mvn) [reply sqsdomain/cw_pool.go:62]: Got it, let's leave it as is. Thank you!

**Pull request merged**
